common/djangoapps/contentserver/middleware.py

```diff
@@ -36,14 +39,24 @@ def is_asset_request(self, request):
             request.path.startswith('/' + XASSET_LOCATION_TAG + '/')
             or
             request.path.startswith('/' + AssetLocator.CANONICAL_NAMESPACE)
+            or
+            self.is_versioned_request(request)
         )
 
+    def is_versioned_request(self, request):
+        return request.path.startswith(VERSIONED_ASSETS_PREFIX)
+
     def process_request(self, request):
         """Process the given request"""
         if self.is_asset_request(request):
             # Make sure we can convert this request into a location.
             if AssetLocator.CANONICAL_NAMESPACE in request.path:
                 request.path = request.path.replace('block/', 'block@', 1)
+
+            # If this is a versioned request, chop off the prefix.
+            if self.is_versioned_request(request):
+                request.path = re.sub(VERSIONED_ASSETS_PATTERN, '', request.path)
```

Comment:
I get where you're coming from, but I think this might be a weird situation that isn't possible and/or relevant.

Let's assume that either a CDN or end user makes request, for the first time, to a given file.  The link they use hash a hash of ABC.  The file may or may not be version ABC when they request it, but they get something back.  They're told to cache this file for a long time.

Now, the asset is updated by an author, which in turn changes its hash.  The links being generated now have a hash of DEF.  The end-user requests DEF, the CDN doesn't have it, and so also requests DEF from origin.

The cache lifetime is so long that a CDN or browser should not be trying to revalidate it.  The only situation I can see here that might be possible is that, somehow, the page still tells the browser to request the ABC version, but the file is really DEF behind the scenes.  It's hard to think of a situation where this might happen.

If it did happen, though, we'd be in a weird place where we would constantly be sending back a 303 redirect when ABC was requested.  We'd be at the behest of the original page being invalidated and starting to use a new link to DEF.  I'm almost positive the browser wouldn't be able to discern that, when requesting ABC, it should follow the redirect and use the content of DEF in place of ABC.  That would also be precisely the case you want to avoid -- hash/content mismatch.

I might be overlooking something here.  I think the fact we use the hash of the current file in assetstore means we can't ever be out-of-sync.  If we switch to a new hash, by definition, the new file is there and the first time a hash is requested, we'll get the latest data.  We would need to encounter a hash collision to have an improperly-cached version.

Does that make sense?